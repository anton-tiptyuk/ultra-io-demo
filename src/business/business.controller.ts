import { Controller, Get } from '@nestjs/common';

import { BusinessService } from './business.service';

@Controller('business')
export class BusinessController {
  constructor(private readonly service: BusinessService) { }

  @Get('trigger-aging')
  async triggerAging() {
    await this.service.performAging();
    return 'aging performed';
  }
}
