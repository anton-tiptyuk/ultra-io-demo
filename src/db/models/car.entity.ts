import {
  Entity,
  PrimaryColumn,
  Column,
  ManyToOne,
  OneToMany,
} from 'typeorm';

import { decimalTransformer } from '../../common/util';

import { Manufacturer } from './manufacturer.entity';
import { CarToOwner } from './carToOwner.entity';

@Entity()
// @Index('uq_box_title_lower', { synchronize: false })
export class Car {
  @PrimaryColumn({ length: 40 })
  id: string;

  @ManyToOne(() => Manufacturer, { onDelete: 'CASCADE', nullable: false })
  manufacturer: Manufacturer;

  @Column('decimal', { precision: 10, scale: 2, transformer: decimalTransformer })
  price: number;

  @Column()
  firstRegistrationDate: Date;

  @OneToMany(() => CarToOwner, carToOwners => carToOwners.car)
  carToOwners: CarToOwner[];
}
