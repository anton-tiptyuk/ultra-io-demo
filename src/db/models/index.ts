export * from './owner.entity';
export * from './manufacturer.entity';
export * from './car.entity';