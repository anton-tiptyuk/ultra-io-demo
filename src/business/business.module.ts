import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { Manufacturer, Owner, CarToOwner } from '../db/models';

import { BusinessController } from './business.controller';
import { BusinessService } from './business.service';

@Module({
  imports: [TypeOrmModule.forFeature([Manufacturer, Owner, CarToOwner])],
  controllers: [BusinessController],
  providers: [BusinessService],
})
export class BusinessModule { }
