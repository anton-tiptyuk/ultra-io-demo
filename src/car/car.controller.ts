import { Controller, Get, Param } from '@nestjs/common';
import { Crud, CrudController } from '@nestjsx/crud';

import { Car } from '../db/models';
import { CarService } from './car.service';

@Crud({
  model: { type: Car },
  query: {
    join: {
      manufacturer: { eager: true },
    },
  },
  params: {
    id: {
      // type: 'uuid',
      // though i seed UIDs for all entitites, the task says ID fields should be string
      type: 'string',
      primary: true,
      field: 'id',
    },
  },
  routes: { exclude: ['createManyBase', 'replaceOneBase'] },
})
@Controller('car')
export class CarController implements CrudController<Car> {
  constructor(public service: CarService) { }

  @Get(':id/manufacturer')
  getManufacturer(@Param('id') id: string) {
    return this.service.getManufacturer(id);
  }
}
