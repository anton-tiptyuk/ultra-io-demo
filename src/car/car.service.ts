import { Injectable } from '@nestjs/common';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';

import { Car } from '../db/models';

@Injectable()
export class CarService extends TypeOrmCrudService<Car> {
  constructor(@InjectRepository(Car) repo: Repository<Car>) {
    super(repo);
  }

  decidePagination() {
    return false;
  }
}
