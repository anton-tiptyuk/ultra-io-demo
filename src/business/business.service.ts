import { Repository } from 'typeorm';
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';

import { Owner } from '../db/models';
import { CarService } from '../car/car.service';

@Injectable()
export class BusinessService {
  constructor(
    private readonly carService: CarService,
    @InjectRepository(Owner)
    private readonly ownerRepo: Repository<Owner>
  ) { }

  // First we call carService
  // It applies the car price with discount
  // and drops CarToOwner join table entitites with purchaseDate matching range

  // Then we get back here and drop orphan owners
  // (because even though we seed data in a way owner posesses only one car,
  // the data structure and common logic says it is possible that owner had owned several cars
  // in the past)
  async performAging() {
    const results: { affected?: number | null }[] = await this.carService.performAging();

    results.push(
      await this
        .ownerRepo
        .createQueryBuilder()
        .delete()
        .where('not exists (select 1 from car_to_owner where car_to_owner.owner_id = owner.id)')
        .execute()
    );

    return results.map(({ affected }) => affected);
  }

}
