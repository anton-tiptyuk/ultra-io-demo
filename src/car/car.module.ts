import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { CarController } from './car.controller';
import { CarService } from './car.service';
import { Car, CarToOwner } from '../db/models';

@Module({
  imports: [TypeOrmModule.forFeature([Car, CarToOwner])],
  controllers: [CarController],
  providers: [CarService],
  exports: [CarService],
})
export class CarModule { }
