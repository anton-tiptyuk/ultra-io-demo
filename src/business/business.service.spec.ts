import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';

import { Owner } from '../db/models';
import { CarService } from '../car/car.service';

import { BusinessService } from './business.service';

const carServiceFake = { performAging: () => undefined };
const ownerRepoWhereFake = { where: () => undefined };
let service: BusinessService;

describe('BusinessService', () => {

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        BusinessService,
        {
          provide: CarService,
          useValue: carServiceFake,
        },
        {
          provide: getRepositoryToken(Owner),
          useValue: {
            createQueryBuilder: () => ({
              delete: () => ownerRepoWhereFake
            })
          },
        },
      ],
    }).compile();

    service = module.get(BusinessService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should perform aging', async () => {
    const whereSpy = jest.spyOn(ownerRepoWhereFake, 'where');
    whereSpy.mockImplementation(() => ({ execute: async () => ({ affected: 13 }) }));

    const performAgingSpy = jest.spyOn(carServiceFake, 'performAging');
    performAgingSpy.mockImplementation(async () => [{ affected: 8 }, { affected: 2 }]);

    await expect(service.performAging()).resolves.toMatchSnapshot();

    expect(whereSpy).toHaveBeenCalled();
    expect(whereSpy.mock.calls).toMatchSnapshot();

    expect(performAgingSpy).toHaveBeenCalled();
  });

});
