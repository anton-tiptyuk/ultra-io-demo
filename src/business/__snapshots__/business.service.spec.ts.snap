// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessService should perform aging 1`] = `
Array [
  8,
  2,
  13,
]
`;

exports[`BusinessService should perform aging 2`] = `
Array [
  Array [
    "not exists (select 1 from car_to_owner where car_to_owner.owner_id = owner.id)",
  ],
]
`;
